core-java/src/main/java/org/spine3/server/saga/SagaRepository.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2015, TeamDev Ltd. All rights reserved.
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.spine3.server.saga;
+
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.Multimap;
+import com.google.protobuf.Message;
+import org.spine3.base.CommandContext;
+import org.spine3.base.EventContext;
+import org.spine3.base.EventRecord;
+import org.spine3.server.EntityRepository;
+import org.spine3.server.MultiHandler;
+
+import javax.annotation.Nonnull;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.Set;
+
+import static com.google.common.base.Throwables.propagate;
+
+/**
+ * The abstract base for repositories managing Sagas.
+ *
+ * @see Saga
+ * @author Alexander Litus
+ */
+public abstract class SagaRepository<I, S extends Saga<I, M>, M extends Message>
+        extends EntityRepository<I, S, M> implements MultiHandler {
+
+    /**
+     * The name of the method used for dispatching commands to sagas.
+     *
+     * <p>This constant is used for obtaining {@code Method} instance via reflection.
+     *
+     * @see #dispatchCommand(Message, CommandContext)
+     */
+    private static final String COMMAND_DISPATCHER_METHOD_NAME = "dispatchCommand";
+
+    /**
+     * The name of the method used for dispatching events to sagas.
+     *
+     * <p>This constant is used for obtaining {@code Method} instance via reflection.
+     *
+     * @see #dispatchEvent(Message, EventContext)
+     */
+    private static final String EVENT_DISPATCHER_METHOD_NAME = "dispatchEvent";
+
+    /**
+     * Returns a saga ID based on the information from the command and command context.
+     *
+     * @param command command which the saga handles
+     * @param context context of the command
+     * @return a saga ID
+     */
+    protected abstract I getSagaIdOnCommand(Message command, CommandContext context);
```

Comment:
How about `getId(Message, CommandContext)`?
Reasons:
1. We know that this is a command because we have CommandContext as the second parameter.
2. We have this method as `protected` and asking for an ID assumes we get the ID of things managed by the repository (which is ProcessManager) in this case.